baseplate/observers/prometheus.py

```diff
@@ -0,0 +1,123 @@
+import logging
+import time
+
+from typing import Any
+from typing import Dict
+from typing import Optional
+
+from baseplate import _ExcInfo
+from baseplate import BaseplateObserver
+from baseplate import LocalSpan
+from baseplate import RequestContext
+from baseplate import Span
+from baseplate import SpanObserver
+from baseplate.lib.prometheus_metrics import PrometheusThriftServerMetrics
+
+
+NANOSECONDS_PER_SECOND = 1e9
+
+logger = logging.getLogger(__name__)
+
+
+class PrometheusBaseplateObserver(BaseplateObserver):
+    """Metrics collecting observer.
+
+    This observer reports Prometheus metrics for HTTP or Thrift client and servers.
+
+    There must be a "protocol" tag set on the server_span that indictes the protocol, either "http" or "thrift",
+    i.e. `span.set_tag("protocol", "http")` or `span.set_tag("protocol", "thrift")`.
+
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def on_server_span_created(self, context: RequestContext, server_span: Span) -> None:
+        observer = PrometheusServerSpanObserver()
+        server_span.register(observer)
+
+
+class PrometheusServerSpanObserver(SpanObserver):
+    def __init__(self) -> None:
+        self.tags: Dict[str, Any] = {}
+        self.start_time: Any = None
+        self.metrics: Any = None
+
+    def on_set_tag(self, key: str, value: Any) -> None:
+        self.tags[key] = value
+
+    def get_prefix(self) -> str:
+        return f"{self.protocol}_server"
+
+    def get_labels(self) -> Dict[str, str]:
+        return self.tags
+
+    @property
+    def protocol(self) -> str:
+        return self.tags.get("protocol", "unknown")
+
+    def set_metrics_by_protocol(self) -> None:
```

Comment:
I've been staring myself a bit blind trying to figure this out, so someone more familiar with baseplate should verify this. I think we're breaking the abstraction here. We should implement a Prometheus Span observer, but it should only be used for shipping the results of span traces to prometheus. The actual metrics specific to thrift or http should either be implemented as spans in the server (e.g. making a child span in the thrift server when handling a request), or - my favorite - just directly in the thrift or http server.

Reply:
i like this idea. i would love to implement this differently. i like the idea of doing the metrics directly in the server. less discuss more